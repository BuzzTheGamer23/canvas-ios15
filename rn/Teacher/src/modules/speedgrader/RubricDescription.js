//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import { View, StyleSheet, Text, ScrollView } from 'react-native'
import { connect } from 'react-redux'
import i18n from 'format-message'
import Navigator from '../../routing/Navigator'
import Screen from '../../routing/Screen'
import WebContainer from '../../common/components/WebContainer'

export class RubricDescription extends Component {

  dismiss = () => {
    this.props.navigator.dismiss()
  }

  renderLongDescription () {
    if (!this.props.description || this.props.description.length === 0) {
      return (
        <View style={styles.emptyState}>
          <Text style={styles.emptyStateText}>{i18n('There currently is no long description for this item.')}</Text>
        </View>
      )
    }
    return (
      <View style={styles.container}>
        <ScrollView bounces={false}>
          <WebContainer html={this.props.description} scrollEnabled={false} navigator={this.props.navigator}/>
        </ScrollView>
      </View>
    )
  }

  render () {
    const description = this.renderLongDescription()
    return (
      <Screen
        title={i18n('Long Description')}
        rightBarButtons={[{
          title: i18n('Done'),
          style: 'done',
          testID: 'rubric-description.done',
          action: this.dismiss,
        }]}>
        { description }
      </Screen>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    paddingTop: 24,
    paddingHorizontal: 16,
  },
  emptyState: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 32,
  },
  emptyStateText: {
    color: '#2D3B45',
    fontSize: 16,
    lineHeight: 19,
    textAlign: 'center',
  },
})

export function mapStateToProps (state: AppState, ownProps: RubricDescriptionOwnProps): RubricDescriptionDataProps {
  let rubric = state.entities.assignments[ownProps.assignmentID].data.rubric
  if (!rubric) {
    return { description: '' }
  }

  return {
    description: rubric.find(r => r.id === ownProps.rubricID).long_description,
  }
}

const Connected = connect(mapStateToProps)(RubricDescription)
export default (Connected: any)

type RubricDescriptionOwnProps = {
  assignmentID: string,
  rubricID: string,
  navigator: Navigator,
}

type RubricDescriptionDataProps = {
  description: string,
}
