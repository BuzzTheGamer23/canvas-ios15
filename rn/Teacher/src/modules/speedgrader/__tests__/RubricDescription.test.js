// @flow

import React from 'react'
import { RubricDescription, mapStateToProps } from '../RubricDescription'
import renderer from 'react-test-renderer'

jest.unmock('ScrollView')

const templates = {
  ...require('../../../__templates__/react-native-navigation'),
  ...require('../../../redux/__templates__/app-state'),
  ...require('../../../api/canvas-api/__templates__/rubric'),
  ...require('../../../api/canvas-api/__templates__/assignments'),
}

let ownProps = {
  assignmentID: '1',
  rubricID: '1',
  navigator: templates.navigator(),
}

let defaultProps = {
  ...ownProps,
  description: 'A satisfactory description',
}

describe('RubricDescription', () => {
  beforeEach(() => jest.resetAllMocks())

  it('properly sets navigator event callback', () => {
    let tree = renderer.create(
      <RubricDescription {...defaultProps} />
    )

    expect(defaultProps.navigator.setOnNavigatorEvent).toHaveBeenCalledWith(tree.getInstance().onNavigatorEvent)
  })

  it('calls dismiss modal when the done button is pressed', () => {
    let tree = renderer.create(
      <RubricDescription {...defaultProps} />
    )

    let instance = tree.getInstance()
    instance.onNavigatorEvent({
      type: 'NavBarButtonPress',
      id: 'done',
    })

    expect(defaultProps.navigator.dismissModal).toHaveBeenCalled()
  })

  it('sets the title of the view', () => {
    renderer.create(
      <RubricDescription {...defaultProps} />
    )

    expect(defaultProps.navigator.setTitle).toHaveBeenCalled()
  })

  it('renders', () => {
    let tree = renderer.create(
      <RubricDescription {...defaultProps} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('mapStateToProps', () => {
  it('returns an empty description when there is no rubric', () => {
    let state = templates.appState({
      entities: {
        assignments: {
          '1': {
            data: templates.assignment({
              id: '1',
            }),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.description).toEqual('')
  })

  it('returns the rubric description whent here is a rubric', () => {
    let rubric = templates.rubric({ id: '1' })
    let state = templates.appState({
      entities: {
        assignments: {
          '1': {
            data: templates.assignment({
              id: '1',
              rubric: [rubric],
            }),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.description).toEqual(rubric.long_description)
  })
})
