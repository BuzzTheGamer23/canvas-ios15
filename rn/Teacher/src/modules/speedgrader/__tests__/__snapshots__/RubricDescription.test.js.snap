// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RubricDescription renders 1`] = `
<View
  style={
    Object {
      "paddingHorizontal": 16,
      "paddingTop": 24,
    }
  }
>
  <RCTScrollView
    DEPRECATED_sendUpdatedChildFrames={false}
    alwaysBounceHorizontal={undefined}
    alwaysBounceVertical={true}
    bounces={false}
    onContentSizeChange={null}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={undefined}
    onResponderTerminationRequest={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onScrollShouldSetResponder={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    scrollEventThrottle={undefined}
    sendMomentumEvents={false}
    style={
      Array [
        Object {
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
          "overflow": "scroll",
        },
        undefined,
      ]
    }
  >
    <RCTScrollContentView
      collapsable={false}
      removeClippedSubviews={undefined}
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <View
        onLayout={[Function]}
        style={undefined}
        testID="web-container.view"
      />
    </RCTScrollContentView>
  </RCTScrollView>
</View>
`;

exports[`RubricDescription renders without description 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "paddingHorizontal": 32,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#2D3B45",
        "fontSize": 16,
        "lineHeight": 19,
        "textAlign": "center",
      }
    }
  >
    There currently is no long description for this item.
  </Text>
</View>
`;
